pkg/chunk/gcp/gcs_object_client.go

```diff
@@ -45,71 +52,111 @@ func NewGCSObjectClient(ctx context.Context, cfg GCSConfig, schemaCfg chunk.Sche
 
-		if _, err := writer.Write(buf); err != nil {
-			return err
-		}
-		if err := writer.Close(); err != nil {
+		if err := s.PutObject(ctx, chunk.ExternalKey(), bytes.NewReader(buf)); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func (s *gcsObjectClient) GetChunks(ctx context.Context, input []chunk.Chunk) ([]chunk.Chunk, error) {
+func (s *GCSObjectClient) GetChunks(ctx context.Context, input []chunk.Chunk) ([]chunk.Chunk, error) {
 	return util.GetParallelChunks(ctx, input, s.getChunk)
 }
 
-func (s *gcsObjectClient) getChunk(ctx context.Context, decodeContext *chunk.DecodeContext, input chunk.Chunk) (chunk.Chunk, error) {
+func (s *GCSObjectClient) getChunk(ctx context.Context, decodeContext *chunk.DecodeContext, input chunk.Chunk) (chunk.Chunk, error) {
+	readCloser, err := s.GetObject(ctx, input.ExternalKey())
+	if err != nil {
+		return chunk.Chunk{}, err
+	}
+
+	defer readCloser.Close()
+
+	buf, err := ioutil.ReadAll(readCloser)
+	if err != nil {
+		return chunk.Chunk{}, err
+	}
+
+	if err := input.Decode(decodeContext, buf); err != nil {
+		return chunk.Chunk{}, err
+	}
+
+	return input, nil
+}
+
+// Get object from the store
+func (s *GCSObjectClient) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, error) {
 	if s.cfg.RequestTimeout > 0 {
 		// The context will be cancelled with the timeout or when the parent context is cancelled, whichever occurs first.
 		var cancel context.CancelFunc
 		ctx, cancel = context.WithTimeout(ctx, s.cfg.RequestTimeout)
 		defer cancel()
 	}
 
-	reader, err := s.bucket.Object(input.ExternalKey()).NewReader(ctx)
-	if err != nil {
-		return chunk.Chunk{}, errors.WithStack(err)
-	}
-	defer reader.Close()
+	return s.bucket.Object(objectKey).NewReader(ctx)
+}
 
-	buf, err := ioutil.ReadAll(reader)
-	if err != nil {
-		return chunk.Chunk{}, errors.WithStack(err)
+// Put object into the store
+func (s *GCSObjectClient) PutObject(ctx context.Context, objectKey string, object io.ReadSeeker) error {
+	writer := s.bucket.Object(objectKey).NewWriter(ctx)
+	// Default GCSChunkSize is 8M and for each call, 8M is allocated xD
+	// By setting it to 0, we just upload the object in a single a request
+	// which should work for our chunk sizes.
+	writer.ChunkSize = s.cfg.ChunkBufferSize
+
+	if _, err := io.Copy(writer, object); err != nil {
+		return err
+	}
+	if err := writer.Close(); err != nil {
+		return err
 	}
 
-	if err := input.Decode(decodeContext, buf); err != nil {
-		return chunk.Chunk{}, err
+	return nil
+}
+
+// List objects from the store
+func (s *GCSObjectClient) List(ctx context.Context, prefix string) (map[string]time.Time, error) {
+	objectKeysWithMtime := map[string]time.Time{}
+	prefixWithSep := prefix + "/"
+
+	iter := s.bucket.Objects(ctx, &storage.Query{Prefix: prefix})
+	for {
+		attr, err := iter.Next()
+		if err != nil {
+			if err == iterator.Done {
+				break
+			} else {
+				return nil, err
+			}
+		}
+
+		if attr.Name == prefixWithSep {
```

Comment:
Could you explain this, please? If you confirm it's what we want, please add a comment to give some details about why we need it.